src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6950,76 +6994,96 @@ private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
-            // identifier token as a calling convention if the token afterwards is actually a
-            // possible start to the function pointer parameter list. The intuition is that, for
-            // some partial trees like this:
-            //
-            //  (delegate*MyProperty.MyMethod();
-            //
-            // The user is actually in the middle of adding a cast, and if we interpreted MyProperty
-            // as a calling convention, we'd get a much worse set of errors than if we treated the
-            // function pointer as having no calling convention.
-            return IsPossibleFunctionPointerParameterListStart(PeekToken(1));
+                    FunctionPointerUnmanagedCallingConventionListSyntax? unmanagedCallingConventions = null;
+                    if (CurrentToken.Kind == SyntaxKind.OpenBracketToken)
+                    {
+                        var openBracket = EatToken(SyntaxKind.OpenBracketToken);
+                        var callingConventionModifiers = _pool.AllocateSeparated<FunctionPointerUnmanagedCallingConventionSyntax>();
+                        var saveTerm = _termState;
+                        _termState |= TerminatorState.IsEndOfFunctionPointerCallingConvention;
+
+                        try
+                        {
+                            while (true)
+                            {
+                                callingConventionModifiers.Add(SyntaxFactory.FunctionPointerUnmanagedCallingConvention(EatToken(SyntaxKind.IdentifierToken)));
+
+                                if (skipBadFunctionPointerTokens(callingConventionModifiers) == PostSkipAction.Abort)
+                                {
+                                    break;
+                                }
+
+                                Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                                callingConventionModifiers.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                            }
+
+                            var closeBracket = EatToken(SyntaxKind.CloseBracketToken);
+
+                            unmanagedCallingConventions = SyntaxFactory.FunctionPointerUnmanagedCallingConventionList(openBracket, callingConventionModifiers, closeBracket);
+                        }
+                        finally
+                        {
+                            _termState = saveTerm;
+                            _pool.Free(callingConventionModifiers);
+                        }
+                    }
+
+                    if (managedSpecifier.Kind == SyntaxKind.ManagedKeyword && unmanagedCallingConventions != null)
+                    {
+                        // 'managed' calling convention cannot be combined with unmanaged calling convention specifiers.
+                        unmanagedCallingConventions = AddError(unmanagedCallingConventions, ErrorCode.ERR_CannotSpecifyManagedWithUnmanagedSpecifiers);
```

Comment:
>unmanagedCallingConventions [](start = 63, length = 27)

Consider attaching the first token rather than to the entire node. I think we have a helper to do that.

Reply:
I actually intentionally covered the entire node. The first token would just be the `[`, which I don't think would be great. I'd prefer to capture the entire thing.

---
In reply to: [451232665](https://github.com/dotnet/roslyn/pull/45684#discussion_r451232665) [](ancestors = 451232665)

Reply:
> I actually intentionally covered the entire node. The first token would just be the [, which I don't think would be great. I'd prefer to capture the entire thing.

Please consider that keeping smaller error span often leads to a better experience because there is less chances that spans of different errors are going to overlap.

---
In reply to: [451851393](https://github.com/dotnet/roslyn/pull/45684#discussion_r451851393) [](ancestors = 451851393,451232665)